app/web_framework/cluster_master_application.py

```diff
@@ -54,9 +54,11 @@ def __init__(self, cluster_master):
                 RouteNode(r'queue', _QueueHandler),
                 RouteNode(r'slave', _SlavesHandler, 'slaves').add_children([
                     RouteNode(r'(\d+)', _SlaveHandler, 'slave').add_children([
-                        RouteNode(r'shutdown', _SlaveShutdownHandler, 'shutdown')
+                        RouteNode(r'shutdown', _SlaveShutdownHandler, 'shutdown'),
+                        RouteNode(r'heartbeat', _SlavesHeartbeatHandler)
```

Comment:
So.. I actually wanted to add it to v2 only, but seems like slave only uses v1 APIs to talk to master. I was unsure if there was a specific reason for that. I just decided to be consistent with the existing code.

I think we could use the `/slave/<id>` endpoint, especially since we are updating the same data structure. My initial thoughts were to maintain separate data structure to track heartbeat, but that seemed overly complicated when I got to the implementation.

Reply:
It seems like the `/slave/<id>` endpoint currently only handles slave state change functionality. Heartbeat is not really a state change. So, I think we either keep a separate `heartbeat` end point as it is right now, or send heartbeat to `/slave/<id>` with current state of the slave, and add logic in master to do nothing if there is no state change. What do you guys think?

Regardless, I think `_last_heartbeat_time` should be updated for both of those requests.

Reply:
yeah.. I thought about that. I'll not make any changes now, but definitely keep that in mind if we decide to revisit this.